tools/lint.mjs

```diff
@@ -0,0 +1,237 @@
+  error(`Prefer "otherwise" to "else": ${format(match)}`);
+}
+for (const match of text.matchAll(/ not the same as /g)) {
+  error(`Prefer "not equal to": ${format(match)}`);
+}
+
+// Look for incorrect use of shape for an MLOperandDescriptor
+for (const match of source.matchAll(/(\|\w*desc\w*\|)'s \[=MLOperand\/shape=\]/ig)) {
+  error(`Use ${match[1]}.{{MLOperandDescriptor/dimensions}} not shape: ${format(match)}`);
+}
+
+// Look for missing dict-member dfns
+for (const element of root.querySelectorAll('.idl dfn[data-dfn-type=dict-member]')) {
+  error(`Dictionary member missing dfn: ${element.innerText}`);
+}
+
+// Look for [] used in algorithm for anything but issues, indexing, slots, and refs
+for (const element of root.querySelectorAll('.algorithm li p:not(.issue)')) {
+  // Exclude \w[ for indexing (e.g. shape[n])
+  // Exclude [[ for inner slots (e.g. [[name]])
+  // Exclude [A for references (e.g. [WEBIDL])
+  for (const match of element.innerText.matchAll(/(?<!\w|\[|\]|«)\[(?!\[|[A-Z])/g)) {
+    error(`Non-index use of [] in algorithm: ${format(match)}`);
+  }
+}
+
+// Ensure vars are method/algorithm arguments, or initialized correctly
+for (const algorithm of root.querySelectorAll('.algorithm')) {
+  const vars = algorithm.querySelectorAll('var');
+  const seen = new Set();
+  for (const v of vars) {
+    const name = v.innerText.trim().replaceAll(/\s+/g, ' ');
+
+    if (v.parentNode.tagName === 'CODE' && v.parentNode.parentNode.tagName === 'DFN') {
+      // Argument definition for IDL method algorithms
+      // e.g. "The compute(graph, inputs, outputs) method steps are:"
+      seen.add(name);
+    } else if (v.parentNode.querySelectorAll('dfn').length) {
+      // Argument definition for abstract algorithms
+      // e.g. "To execute graph, given MLGraph graph ..."
+      seen.add(name);
+    } else {
+      const text = v.parentNode.innerText.trim().replaceAll(/\s+/g, ' ');
+      const patterns = [
+        // "Let var be ..."
+        // "Let var1 be ... and and var2 be ..."
+        'let( .* and)? ' + name + ' be',
+
+        // "Let var given ... be ..." (for lambdas)
+        'let ' + name + ' given .* be',
+
+        // "For each var ..."
+        // "For each ...  → var ..."
+        'for each( \\w+ →)? ' + name,
+      ];
+      if (patterns.some(p => new RegExp('\\b' + p + '\\b', 'i').test(text))) {
+        // Variable declaration/initialization
+        seen.add(name);
+      } else if (new RegExp('\\bgiven .* \\b' + name + '\\b', 'i').test(text)) {
+        // Lambda argument declarations
+        // e.g. "Let validationSteps given MLOperandDescriptor descriptor be..."
+        seen.add(name);
+      } else if (!seen.has(name)) {
+        error(`Uninitialized variable "${name}" in "${algorithm.getAttribute('data-algorithm')}": ${text}`);
+        seen.add(name);
+      }
+    }
+  }
+}
+
+// Prevent accidental normative references to other specs
+const NORMATIVE_REFERENCES = new Set([
+  '[ECMASCRIPT]',
```

Comment:
It's the opposite - this stops us from accidentally adding a reference. So if you write `[=transferred=]` somewhere and Bikeshed decides that references a Credentials spec, this would flag an error unless that spec is added here. https://github.com/webmachinelearning/webnn/commit/b369a88aa09e4b4c941ff7c3a88240cccf0fdad3

Reply:
Improved the documentation in 736c54c